### ISSUE TITLE
PEP 695 variance inference for covariant type params

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pyright#7929]
**Describe the bug**
Encountered this one while converting existing TypeVars to the PEP 695 syntax.
```py
from typing import Callable, Generic, TypeVar

class Job[_R_co]:
    def __init__(self, target: Callable[[], _R_co]) -> None:
        self.target = target

def func(
    action: Job[int | None],
    a1: Job[int],
) -> None:
    action = a1  # error
```
```py
  test.py:109:14 - error: Expression of type "Job[int]" is incompatible with declared type "Job[int | None]"
    "Job[int]" is incompatible with "Job[int | None]"
      Type parameter "_R_co@Job" is invariant, but "int" is not the same as "int | None" (reportAssignmentType)
```

In the example above, I'd have expected `_R_co` to be inferred as covariant - the same as a callable return type. Using `Generic` and `TypeVar` with the explicit `covariant=True` works as expected:
```py
_S_co = TypeVar("_S_co", covariant=True)

class Job2(Generic[_S_co]):
    def __init__(self, target: Callable[[], _S_co]) -> None:
        self.target = target

def func2(
    action: Job2[int | None],
    a1: Job2[int],
) -> None:
    action = a1
```

**VS Code extension or command-line**
1.1.363

**Comment by erictraut** [comment]
I think pyright's behavior here is correct, so I don't consider this a bug.

The key here is the instance variable `target` which has an inferred type of `Callable[[], _R_co]`. This is a writable instance variable (not read-only), so its type must be invariant. That means `_R_co` must be invariant as well.

If you comment out the line `self.target = target`, you'll see that the type error disappears because `_R_co` is now inferred to be covariant.

I'll note that neither pyright nor mypy enforce or validate TypeVar variance for non-protocol classes (prior to PEP 695). Both type checkers perform simplistic checks for egregious misuse of covariant or contravariant TypeVars in parameters or return types, but these checks are far from comprehensive. With that in mind, the fact that using `covariant=True` with the old syntax doesn't mean that it was previously correct. I find that developers often define TypeVar variance incorrectly, which is one of the reasons I pushed for automatic invariance inference in PEP 695.

**Comment by cdce8p** [comment]
> The key here is the instance variable `target` which has an inferred type of `Callable[[], _R_co]`. This is a writable instance variable (not read-only), so its type must be invariant. That means `_R_co` must be invariant as well.

So making it `Final` would preserve the covariance then? That seems to work:

```py
from typing import Callable, Final

class Job[_R_co]:
    def __init__(self, target: Callable[[], _R_co]) -> None:
        self.target: Final = target

def func(
    action: Job[int | None],
    a1: Job[int],
) -> None:
    action = a1
```
